compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -3869,28 +3876,71 @@ public void visit(BLangOnFailClause onFailClause, AnalyzerData data) {
         // Create a new block environment for the on-fail node.
         SymbolEnv onFailEnv = SymbolEnv.createBlockEnv(onFailClause.body, data.env);
         VariableDefinitionNode onFailVarDefNode = onFailClause.variableDefinitionNode;
+        Stack<LinkedHashSet<BType>> onFailErrTypes = data.commonAnalyzerData.errorTypes;
+
         if (onFailVarDefNode != null) {
+            BLangVariable variableNode = (BLangVariable) onFailVarDefNode.getVariable();
+            if (variableNode.getKind() != NodeKind.VARIABLE && variableNode.getKind() != NodeKind.ERROR_VARIABLE) {
+                dlog.error(variableNode.pos, DiagnosticErrorCode.INVALID_BINDING_PATTERN_IN_ON_FAIL);
+            }
+
+            BType failErrorType;
+            LinkedHashSet<BType> onFailErrType = onFailErrTypes.peek();
+            boolean isVarTypeHardcoded = false;
+            if (onFailErrType.size() == 1) {
+                failErrorType = onFailErrType.iterator().next();
+            } else if (onFailErrType.size() > 1) {
+                failErrorType = BUnionType.create(null, onFailErrType);
+            } else {
+                failErrorType = symTable.errorType;
+                isVarTypeHardcoded = true;
```

Comment:
Why do we need to special case like this? Can't we use something like noType as the type instead to handle this? Or never.